Tutorial about using event handlers and substitutions in launch files

kurshakuz opened the pull request:
This tutorial covers usage examples of using event handlers and substations in launch files.

It is an advanced tutorial, but placed in that particular place as it made more sense to follow from one launch tutorial to another. 

I think it would be even better to create a folder for launch tutorials as it is already made with TF2 and URDF. Let me know if I should make these updates in that PR or create a new one if you agree.

Signed-off-by: Shyngyskhan Abilkassov abilkasov@gmail.com

kurshakuz committed:
create a tutorial about using event handlers and substitutions in launch files

kurshakuz commented:
> I'd say let's do it in a separate PR, that way this one can just concentrate on one thing.

Great! I will then create a new PR after that one is merged.

audrow commented on source/Tutorials/Launch-Files/Using-Event-Handlers-and-Substitutions.rst:
```diff
@@ -0,0 +1,482 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the ``example.launch.py`` file, ``turtlesim_ns``, ``use_provided_red``, and ``new_background_r`` launch configurations and arguments are defined.
+``LaunchConfiguration`` substitution allows to acquire the value of the launch argument in any part of the launch description.
```
This seems to be worded oddly to me. Maybe the following?

```suggestion
These ``LaunchConfiguration`` substitutions allow us to acquire the value of the launch argument in any part of the launch description.
```

kurshakuz replied:
https://github.com/ros2/ros2_documentation/pull/2094/commits/1bbf45a8cd587b2f06fd1c6cd958728cc528214d

audrow commented on source/Tutorials/Launch-Files/Using-Event-Handlers-and-Substitutions.rst:
```diff
@@ -0,0 +1,482 @@
+        name='sim'
+    )
+
+Afterwards, the ``ExecuteProcess`` action called ``spawn_turtle`` is defined with the corresponding ``cmd`` argument.
+It makes a call to the spawn service of the turtlesim node.
```
```suggestion
Afterwards, the ``ExecuteProcess`` action called ``spawn_turtle`` is defined with the corresponding ``cmd`` argument and it makes a call to the spawn service of the turtlesim node.
```

kurshakuz replied:
https://github.com/ros2/ros2_documentation/pull/2094/commits/1bbf45a8cd587b2f06fd1c6cd958728cc528214d

audrow reviewed (changes requested):
This is great content overall, especially the event handler information. I have two larger comments for this PR:
1. Should this be split into two PRs? It seems like event handlers and substitutions both have enough content to be discussed separately.
2. In the examples given, I'm not sure why substitutions are necessary. I'm not positive, but it seems like similar functionality could be achieved in most cases without substitutions. It might be good to focus on cases where substitution is the only solution, or maybe to explain why these examples are not possible without substitutions.

Lastly, it would be good to have a summary section at the end that reiterates why what they have just learned is useful, as you did in the background.

kurshakuz commented:
> 1. Should this be split into two PRs? It seems like event handlers and substitutions both have enough content to be discussed separately.

That sounds reasonable. I initially made two different launch file examples, but they had some major overlaps and that was the reason for combining them. 

What I think I could is to split this one into two separate tutorials using the same example code. One about using substitutions and the next one extending the previous one with event handlers. What do you think?

> 2. In the examples given, I'm not sure why substitutions are necessary. I'm not positive, but it seems like similar functionality could be achieved in most cases without substitutions. It might be good to focus on cases where substitution is the only solution, or maybe to explain why these examples are not possible without substitutions.

This! I really had to come up with some superficial examples of substitutions just to mention them and show their possible usage examples. The problem with substitutions is that they either not widely used or used for a specific cases, like `urdf` loading via `xacro` command. On the other hand, my goal was to make the tutorial example easy replicate and understand (without using external packages and simulation tools), and to make sure that reader gets acquainted with capabilities of substitutions. I would love to hear if you or someone else could share better examples of using substitutions!

kurshakuz committed:
apply review suggestions

kurshakuz committed:
split subsitution and even hanlder tutorials

kurshakuz committed:
add summary sections and internal links

kurshakuz committed:
remove FindExecutable usage examples

kurshakuz committed:
fix grammar

kurshakuz committed:
remove old combined tutorial

kurshakuz committed:
Merge pull request #8 from kurshakuz/advanced-launch-divided

Split up substitutions and event handlers tutorials

kurshakuz commented:
I finished the requested updates. Now there are two separate tutorials covering substitutions and event handlers. I also removed some unnecessary use cases of Substitutions, like the `FindExecuable` which was not making a lot of sense in that situation. Let me know if something more could still be improved, removed, or added.

Summaries and corresponding links to new tutorials are added.

ahcorde commented on source/Tutorials/Launch-Files/Using-Event-Handlers.rst:
```diff
@@ -0,0 +1,298 @@
+
+Create a new file called ``example_event_handlers.launch.py`` file in the ``/launch`` folder of the ``launch_tutorial`` package.
+
+.. code-block:: python
```
is this example living in a repository ?

ahcorde commented on source/Tutorials/Launch-Files/Using-Substitutions.rst:
```diff
@@ -0,0 +1,331 @@
+Firstly, we will create a launch file that will call and pass arguments to another launch file.
+To do this, create an ``example_main.launch.py`` file in the ``/launch`` folder of the ``launch_tutorial`` package.
+
+.. code-block:: python
```
some question here, are these two python files living somewhere?

kurshakuz replied:
No, they are not hosted anywhere. Do you think I should add them?

They don't require to be a part of a package and I thought they also don't bring a lot of value without a tutorial context. I did not add files from the [Using ROS2 launch for large projects](https://docs.ros.org/en/rolling/Tutorials/Launch-Files/Using-ROS2-Launch-For-Large-Projects.html) tutorial for the same reason.

audrow commented on source/Tutorials/Launch-Files/Using-Substitutions.rst:
```diff
@@ -0,0 +1,331 @@
+        ])
+
+In the ``example_substitutions.launch.py`` file, ``turtlesim_ns``, ``use_provided_red``, and ``new_background_r`` launch configurations and arguments are defined.
+These ``LaunchConfiguration`` substitutions allow us to acquire the value of the launch argument in any part of the launch description.
```
It would be good to say what this behavior is if you do not use `LaunchConfiguration`.

kurshakuz replied:
I am not totally sure what you mean by that. Do you want to say that we would not be able to use values of the launch arguments?

I added some additional notes in https://github.com/ros2/ros2_documentation/pull/2094/commits/4f64b44386c2eec6a58c2f5220fc1f80fa1ec8bd, let me know if makes more sense

audrow replied:
Oops, I think I meant, "if you do not use substitutions." This is related to the comment about `TextSubstitution('200')` where I don't understand why you are using substitutions. In other words, what are the `TextSubstitutions` allowing you to do that you wouldn't get without them? At the moment, it's not clear to me. Specifically, how is
```python
        use_provided_red_launch_arg = DeclareLaunchArgument(
            'use_provided_red',
            default_value=TextSubstitution(text='False')
        )
```
different than the following?
```python
        use_provided_red_launch_arg = DeclareLaunchArgument(
            'use_provided_red',
            default_value='False'
        )
```

kurshakuz replied:
Yes, there is no need to use substitutions there. I removed them in the https://github.com/ros2/ros2_documentation/pull/2094/commits/6744cde7fad4848c6dcabfae676b17dc333da5f2

audrow commented on source/Tutorials/Launch-Files/Using-Substitutions.rst:
```diff
@@ -0,0 +1,331 @@
+    )
+    new_background_r_launch_arg = DeclareLaunchArgument(
+        'new_background_r',
+        default_value=TextSubstitution(text='200')
```
Could you just put '200' here and avoid the TextSubstitution? If not, why not? If this is necessary for the logic below, it'd be good to mention this.

kurshakuz replied:
Removed in the https://github.com/ros2/ros2_documentation/pull/2094/commits/faab34edcf641637560682e88089cab620df84d0

audrow reviewed (changes requested):
Looks pretty good. There's a few things with the substitutions that aren't clear to me.

kurshakuz committed:
replace textsubstitution

kurshakuz committed:
add some notes on LaunchConfiguration

audrow commented on source/Tutorials/Launch-Files/Using-Event-Handlers.rst:
```diff
@@ -0,0 +1,298 @@
+        )
+        new_background_r_launch_arg = DeclareLaunchArgument(
+            'new_background_r',
+            default_value=TextSubstitution(text='200')
```
Is this different than just setting the default value to `'200'`?

kurshakuz replied:
You are right, I fixed it in the https://github.com/ros2/ros2_documentation/pull/2094/commits/6744cde7fad4848c6dcabfae676b17dc333da5f2

audrow reviewed (changes requested):
I'm still not exactly sure what substitutions do here that you could not do without them. It would be nice to make this very clear.

kurshakuz committed:
remove textsubstitution usage

kurshakuz committed:
remove unused comma

audrow reviewed (approved):
Looks good to me!

ahcorde commented:
@Mergifyio backport galactic foxy

ros-discourse commented:
This pull request has been mentioned on **Open Robotics Discourse**. There might be relevant details there:

https://discourse.openrobotics.org/t/better-launch-a-replacement-for-the-ros2-launch-system-intuitive-simple-memorable/49348/19
